### PR TITLE
[enocean] Added support for Heat Recovery Ventilation devices (#9465)

### LINKED ISSUE #9465: [enocean] Add support for Heat Recovery Ventilation devices
Heat Recovery Ventilation devices use EEP family D2-50. These EEPs should also be implemented to support these devices.

### PR DESCRIPTION
**Pull request by fruggy83**
* Implemented EEP family D2_50
 * Added new thing type (heatRecoveryVentilation) and channels for heat recovery ventilation units

Fixes #9465

Signed-off-by: Daniel Weber <uni@fruggy.de>

### CODE REVIEW

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java [hunk 1]
```diff
@@ -78,13 +79,17 @@
...
-            THING_TYPE_PUSHBUTTON, THING_TYPE_ROCKERSWITCH, THING_TYPE_CLASSICDEVICE, THING_TYPE_CENTRALCOMMAND,
-            THING_TYPE_ROOMOPERATINGPANEL, THING_TYPE_MECHANICALHANDLE, THING_TYPE_CONTACT,
-            THING_TYPE_MEASUREMENTSWITCH, THING_TYPE_TEMPERATURESENSOR, THING_TYPE_TEMPERATUREHUMIDITYSENSOR,
-            THING_TYPE_GENERICTHING, THING_TYPE_ROLLERSHUTTER, THING_TYPE_OCCUPANCYSENSOR,
-            THING_TYPE_LIGHTTEMPERATUREOCCUPANCYSENSOR, THING_TYPE_LIGHTSENSOR, THING_TYPE_ENVIRONMENTALSENSOR,
-            THING_TYPE_AUTOMATEDMETERSENSOR, THING_TYPE_THERMOSTAT, THING_TYPE_MULTFUNCTIONSMOKEDETECTOR));
+    public static final ThingTypeUID THING_TYPE_HEATRECOVERYVENTILATION = new ThingTypeUID(BINDING_ID,
+            "heatRecoveryVentilation");
+
+    public static final Set<ThingTypeUID> SUPPORTED_DEVICE_THING_TYPES_UIDS = new HashSet<>(
```
Comment (J-N-K) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java:85]:
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_DEVICE_THING_TYPES_UIDS = Set.of(
```

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java [hunk 2]
```diff
@@ -179,6 +184,30 @@
...
+    public static final String CHANNEL_OUTDOORAIRTEMPERATURE = "outdoorAirTemperature";
+    public static final String CHANNEL_SUPPLYAIRTEMPERATURE = "supplyAirTemperature";
+    public static final String CHANNEL_INDOORAIRTEMPERATURE = "indoorAirTemperature";
+    public static final String CHANNEL_EXHAUSTAIRTEMPERATURE = "exhaustAirTemperature";
+    public static final String CHANNEL_SUPPLYAIRFANAIRFLOWRATE = "supplyAirFanAirFlowRate";
+    public static final String CHANNEL_EXHAUSTAIRFANAIRFLOWRATE = "exhaustAirFanAirFlowRate";
+    public static final String CHANNEL_SUPPLYFANSPEED = "supplyFanSpeed";
+    public static final String CHANNEL_EXHAUSTFANSPEED = "exhaustFanSpeed";
+
     public static final Map<String, EnOceanChannelDescription> CHANNELID2CHANNELDESCRIPTION = Collections
```
Comment (J-N-K) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java:211]: Use `Map.of` here

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            case CHANNEL_FIREPLACESAFETYMODE:
+                return getBit(bytes[1], 3) ? OnOffType.ON : OnOffType.OFF;
+            case CHANNEL_HEATEXCHANGERBYPASSSTATUS:
+                return getBit(bytes[1], 2) ? OpenClosedType.OPEN : OpenClosedType.CLOSED;
+            case CHANNEL_SUPPLYAIRFLAPSTATUS:
+                return getBit(bytes[1], 1) ? OpenClosedType.OPEN : OpenClosedType.CLOSED;
+            case CHANNEL_EXHAUSTAIRFLAPSTATUS:
+                return getBit(bytes[1], 0) ? OpenClosedType.OPEN : OpenClosedType.CLOSED;
+            case CHANNEL_DEFROSTMODE:
+                return getBit(bytes[2], 7) ? OnOffType.ON : OnOffType.OFF;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java:123]:
```suggestion
                return OnOffType.from(getBit(bytes[2], 7));
```

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/handler/EnOceanBaseActuatorHandler.java [hunk 4]
```diff
@@ -50,8 +50,9 @@
 public class EnOceanBaseActuatorHandler extends EnOceanBaseSensorHandler {
 
     // List of thing types which support sending of eep messages
-    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = new HashSet<>(Arrays.asList(THING_TYPE_CENTRALCOMMAND,
-            THING_TYPE_MEASUREMENTSWITCH, THING_TYPE_GENERICTHING, THING_TYPE_ROLLERSHUTTER, THING_TYPE_THERMOSTAT));
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = new HashSet<>(
+            Arrays.asList(THING_TYPE_CENTRALCOMMAND, THING_TYPE_MEASUREMENTSWITCH, THING_TYPE_GENERICTHING,
+                    THING_TYPE_ROLLERSHUTTER, THING_TYPE_THERMOSTAT, THING_TYPE_HEATRECOVERYVENTILATION));
```
Comment (J-N-K) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/handler/EnOceanBaseActuatorHandler.java:55]:
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Set.of(THING_TYPE_CENTRALCOMMAND, THING_TYPE_MEASUREMENTSWITCH, THING_TYPE_GENERICTHING,
                    THING_TYPE_ROLLERSHUTTER, THING_TYPE_THERMOSTAT, THING_TYPE_HEATRECOVERYVENTILATION);
```

**Review by J-N-K — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java (hunk 2 above)
Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java]: As I have more than 10 entries here, I am using Map.ofEntries, hope this is the expected way

**Comment by fruggy83** [comment]
@J-N-K Thanks a lot for your review. Your comments also revealed an error in my vscode settings 👍 IDE said JDK11 is used but my settings pointed to JDK8 path. So I could not use these "of" functions. 

As every year: How do we proceed with the license header of the newly added files? Should I already update it to 2021 (Jenkis build will fail) or do you update it during the merge?

**Comment by fwolter** [comment]
You can rebase your branch, then the headers you didn't touch are updated automatically. If you added new code files, you can run `mvn license:format` in the root directoy of this repo to update the years. This should be done before merging.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    protected final byte rmtMask = (byte) 0x0f;
+    protected final byte RMT_BASIC_STATUS = 0x00;
+    protected final byte RMT_EXTENDED_STATUS = 0x01; // not yet implemented
+
+    protected final byte DOMC_NOACTION = 0x0f;
+    protected final byte CONTROL_NOACTION = 0;
+    protected final byte TMOC_NOACTION = 127;
+    protected final byte TMOC_ACTIVATE = (byte) 0xff;
+    protected final byte THRESHOLD_NOACTION = 127;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java:56]: If the fields are immutable and not defined in the constructor, please them static as well.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+
+/**
+ *
+ * @author Daniel Weber - Initial contribution
+ */
+public class D2_50 extends _VLDMessage {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java:40]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_00.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.openhab.binding.enocean.internal.eep.D2_50;
+
+import org.openhab.binding.enocean.internal.messages.ERP1Message;
+
+/**
+ *
+ * @author Daniel Weber - Initial contribution
+ */
+public class D2_50_00 extends D2_50 {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_00.java:21]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_01.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.openhab.binding.enocean.internal.eep.D2_50;
+
+import org.openhab.binding.enocean.internal.messages.ERP1Message;
+
+/**
+ *
+ * @author Daniel Weber - Initial contribution
+ */
+public class D2_50_01 extends D2_50 {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_01.java:21]: Please add `@NonNullByDefault`

And please do so for the rest of the files that aren't dto classes.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_10.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.openhab.binding.enocean.internal.eep.D2_50;
+
+import org.openhab.binding.enocean.internal.messages.ERP1Message;
+
+/**
+ *
+ * @author Daniel Weber - Initial contribution
+ */
+public class D2_50_10 extends D2_50 {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_10.java:21]: What is the point of these classes that don't add any extra functionality? Can you just use the D2_50 class instead?

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java (hunk 5 above)
Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java]: Sorry was my fault, thought Javas final equals c# const. Changed these field to static final.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java (hunk 6 above)
Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java:40]: I would like to add this annotation in another PR to not mess up this PR with non related changes. Adding this annotation would need many changes in other classes, too.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_10.java (hunk 9 above)
Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50_10.java]: D2_50 should have been abstract as this is not an EEP. However all EEPs of D2_50 EEP family use D2_50 class now.

**Comment by fruggy83** [comment]
@cpmeister Thanks a lot for your review comments. I hope I have resolved all problems now. I would like to address this NonNullByDefault point in a seperate PR, if this is ok.

@fwolter Thanks a lot for your hints regarding rebasing.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java (hunk 6 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/D2_50/D2_50.java:40]: understood

**Review by cpmeister — APPROVED** [review]
LGTM

**Pull request merged**